apps/api/src/app/workflows-v2/usecases/build-step-issues/build-step-issues.usecase.ts

```diff
@@ -0,0 +1,258 @@
+import {
+  ContentIssue,
+  JSONSchemaDto,
+  StepContentIssueEnum,
+  StepIssuesDto,
+  UserSessionData,
+  StepTypeEnum,
+  WorkflowOriginEnum,
+  ControlValuesLevelEnum,
+} from '@novu/shared';
+import {
+  InstrumentUsecase,
+  TierRestrictionsValidateUsecase,
+  TierRestrictionsValidateCommand,
+  dashboardSanitizeControlValues,
+  PinoLogger,
+} from '@novu/application-generic';
+
+import _ from 'lodash';
+import Ajv, { ErrorObject } from 'ajv';
+import addFormats from 'ajv-formats';
+import { buildVariables } from '../../util/build-variables';
+import { BuildAvailableVariableSchemaCommand, BuildAvailableVariableSchemaUsecase } from '../build-variable-schema';
+import { BuildStepIssuesCommand } from './build-step-issues.command';
+
+@Injectable()
+export class BuildStepIssuesUsecase {
+  constructor(
+    private buildAvailableVariableSchemaUsecase: BuildAvailableVariableSchemaUsecase,
+    private controlValuesRepository: ControlValuesRepository,
+    private tierRestrictionsValidateUsecase: TierRestrictionsValidateUsecase,
+    private logger: PinoLogger
+  ) {}
+
+  @InstrumentUsecase()
+  async execute(command: BuildStepIssuesCommand): Promise<StepIssuesDto> {
+    const {
+      workflowOrigin,
+      user,
+      step: foundPersistedStep,
+      workflow: persistedWorkflow,
+      stepDto: step,
+      controlSchema,
+    } = command;
+
+    const variableSchema = await this.buildAvailableVariableSchemaUsecase.execute(
+      BuildAvailableVariableSchemaCommand.create({
+        environmentId: user.environmentId,
+        organizationId: user.organizationId,
+        userId: user._id,
+        stepInternalId: foundPersistedStep?.stepId,
+        workflow: persistedWorkflow,
+        ...(step.controlValues ? { optimisticControlValues: step.controlValues } : {}),
+      })
+    );
+
+    let controlValueLocal = step.controlValues;
+
+    if (!controlValueLocal) {
+      controlValueLocal = (
+        await this.controlValuesRepository.findOne({
+          _environmentId: user.environmentId,
+          _organizationId: user.organizationId,
+          _workflowId: persistedWorkflow?._id,
+          _stepId: foundPersistedStep?._templateId,
+          level: ControlValuesLevelEnum.STEP_CONTROLS,
+        })
+      )?.controls;
+    }
+
+    const sanitizedControlValues =
+      controlValueLocal && workflowOrigin === WorkflowOriginEnum.NOVU_CLOUD
+        ? dashboardSanitizeControlValues(this.logger, controlValueLocal, step.type) || {}
+        : this.frameworkSanitizeEmptyStringsToNull(controlValueLocal) || {};
+
+    const controlIssues = this.processControlValuesBySchema(controlSchema, sanitizedControlValues || {});
+    const liquidTemplateIssues = this.processControlValuesByLiquid(variableSchema, controlValueLocal || {});
+    const customIssues = await this.processControlValuesByRules(user, step.type, sanitizedControlValues || {});
+    const customControlIssues = _.isEmpty(customIssues) ? {} : { controls: customIssues };
```

Comment:
Please move this to processControlValuesByRules so that it returns the same interface `StepIssuesDto` as all process functions.

Reply:
BTW ideally it should be `StepIssueDto` to define one step issue and then the return type is the `StepIssueDto[]`

Reply:
Good catch!

we need to remove the StepIssuesDto.body prop as well.